pallets/claims/src/lib.rs

```diff
@@ -0,0 +1,182 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+use codec::Encode;
+use frame_support::{
+	decl_error, decl_event, decl_module, decl_storage,
+	dispatch::DispatchResult,
+	ensure,
+	traits::Get,
+	weights::{DispatchClass, Pays},
+};
+use frame_system::ensure_signed;
+use hex::FromHex;
+use orml_traits::{MultiCurrency, MultiCurrencyExtended};
+use orml_utilities::with_transaction_result;
+use primitives::{Amount, AssetId, Balance, CORE_ASSET_ID};
+use sp_io::{crypto::secp256k1_ecdsa_recover, hashing::keccak_256};
+use sp_runtime::traits::Zero;
+use sp_std::prelude::*;
+use sp_std::vec::Vec;
+
+pub use traits::*;
+
+mod claims_data;
+mod traits;
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+mod benchmarking;
+
+use weights::WeightInfo;
+pub mod weights;
+
+pub trait Config: frame_system::Config {
+	type Event: From<Event<Self>> + Into<<Self as frame_system::Config>::Event>;
+	type Currency: MultiCurrencyExtended<Self::AccountId, CurrencyId = AssetId, Balance = Balance, Amount = Amount>;
+	type Prefix: Get<&'static [u8]>;
+	type WeightInfo: WeightInfo;
+}
+
+type BalanceOf<T> = <<T as Config>::Currency as MultiCurrency<<T as frame_system::Config>::AccountId>>::Balance;
+
+decl_storage! {
+	trait Store for Module<T: Config> as Claims {
+		Claims get(fn claims): map hasher(blake2_128_concat) EthereumAddress => BalanceOf<T>;
+
+		PalletVersion: StorageVersion = StorageVersion::V1EmptyBalances;
+	}
+
+	add_extra_genesis {
+		config(claims): Vec<(EthereumAddress, BalanceOf<T>)>;
+
+		build(|config: &GenesisConfig<T>| {
+			config.claims.iter().for_each(|(eth_address, initial_balance)| {
+				Claims::<T>::mutate(eth_address, |amount| *amount += *initial_balance)
+			})
+		})
+	}
+}
+
+decl_event!(
+	pub enum Event<T>
+	where
+		AccountId = <T as frame_system::Config>::AccountId,
+		Balance = Balance,
+	{
+		Claimed(AccountId, Balance),
+	}
+);
+
+decl_error! {
+	pub enum Error for Module<T: Config> {
+		InvalidEthereumSignature,
+		NoClaimOrAlreadyClaimed,
+	}
+}
+
+decl_module! {
+	pub struct Module<T: Config> for enum Call where origin: T::Origin {
+		type Error = Error<T>;
+
+		fn deposit_event() = default;
+
+		/// The Prefix that is used in signed Ethereum messages for this network
+		const Prefix: &[u8] = T::Prefix::get();
+
+		#[weight = (<T as Config>::WeightInfo::claim(), DispatchClass::Normal, Pays::No)]
+		fn claim(origin, ethereum_signature: EcdsaSignature)  {
```

Comment:
https://github.com/paritytech/substrate/blob/171ad2322d7561934b82f0b7e41c58d1710f8c24/primitives/runtime/src/traits.rs#L777

From: Alexander Popiak
Answering the questions about transaction propagation:
Txs that return error when passed to validate_transaction here are not propagated.

Reply:
cc @enthusiastmartin

Reply:
can be resolved or still a concern?

Reply:
@jak-pan @enthusiastmartin

Reply:
txs validated by SignedExtension

Reply:
handled by the signed extension to validate claim.